### PR TITLE
HBASE-27021 StoreFileInfo should set its initialPath in a consistent way

### PR DESCRIPTION
**Pull request by wchevreuil**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 44s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| -1 :x: |  mvninstall  |   3m 48s |  root in master failed.  |
| +1 :green_heart: |  compile  |   0m 34s |  master passed  |
| +1 :green_heart: |  shadedjars  |   3m 57s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 29s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 13s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 34s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 34s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   3m 55s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   7m 56s |  hbase-server in the patch failed.  |
|  |   |  25m 44s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4419 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux e8eaf504c86b 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f6e9d3e1dd |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| mvninstall | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk8-hadoop3-check/output/branch-mvninstall-root.txt |
| unit | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk8-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/testReport/ |
| Max. process+thread count | 567 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  0s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| -1 :x: |  mvninstall  |   0m 58s |  root in master failed.  |
| -1 :x: |  compile  |   0m 26s |  hbase-server in master failed.  |
| -1 :x: |  shadedjars  |   2m 33s |  branch has 10 errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m  8s |  hbase-server in master failed.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 18s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   4m 42s |  patch has no errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m 31s |  hbase-server generated 96 new + 0 unchanged - 0 fixed = 96 total (was 0)  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  13m 14s |  hbase-server in the patch failed.  |
|  |   |  28m 53s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4419 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 90c6ef53f8a7 5.4.0-1071-aws #76~18.04.1-Ubuntu SMP Mon Mar 28 17:49:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f6e9d3e1dd |
| Default Java | AdoptOpenJDK-11.0.10+9 |
| mvninstall | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/branch-mvninstall-root.txt |
| compile | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/branch-compile-hbase-server.txt |
| shadedjars | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/branch-shadedjars.txt |
| javadoc | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/branch-javadoc-hbase-server.txt |
| javadoc | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/diff-javadoc-javadoc-hbase-server.txt |
| unit | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-jdk11-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/testReport/ |
| Max. process+thread count | 610 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 56s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| -1 :x: |  mvninstall  |   1m  3s |  root in master failed.  |
| -1 :x: |  compile  |   0m 27s |  hbase-server in master failed.  |
| -0 :warning: |  checkstyle  |   0m  4s |  The patch fails to run checkstyle in hbase-server  |
| -1 :x: |  spotless  |   0m 48s |  branch has 66 errors when running spotless:check, run spotless:apply to fix.  |
| -1 :x: |  spotbugs  |   0m  6s |  hbase-server in master failed.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 30s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 38s |  the patch passed  |
| -0 :warning: |  javac  |   2m 38s |  hbase-server generated 193 new + 0 unchanged - 0 fixed = 193 total (was 0)  |
| +1 :green_heart: |  checkstyle  |   0m 34s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  13m 34s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.2 3.3.1.  |
| -1 :x: |  spotless  |   0m 40s |  patch has 66 errors when running spotless:check, run spotless:apply to fix.  |
| +1 :green_heart: |  spotbugs  |   1m 59s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 12s |  The patch does not generate ASF License warnings.  |
|  |   |  31m 18s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4419 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti spotless checkstyle compile |
| uname | Linux 2520e49a53ad 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f6e9d3e1dd |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| mvninstall | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/branch-mvninstall-root.txt |
| compile | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/branch-compile-hbase-server.txt |
| checkstyle | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/buildtool-branch-checkstyle-hbase-server.txt |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/branch-spotless.txt |
| spotbugs | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/branch-spotbugs-hbase-server.txt |
| javac | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/diff-compile-javac-hbase-server.txt |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 64 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java [hunk 1]
```diff
@@ -119,18 +119,18 @@ public class StoreFileInfo implements Configurable {
    */
   public StoreFileInfo(final Configuration conf, final FileSystem fs, final Path initialPath,
     final boolean primaryReplica) throws IOException {
-    this(conf, fs, null, initialPath, primaryReplica);
+    this(conf, fs, fs.getFileStatus(initialPath), primaryReplica);
```
Comment (joshelser) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java:122]: Should make sure constructor javadoc is updated as this is a Namenode call. If we call this often, we're incurring remote reads which could be a bad perf hit. Need to make sure current callers are behaving nicely (we _should_ be caching StoreFileInfo's already)

**Review by joshelser — APPROVED** [review]
I assume you've spot-checked the output of TestHStore after this and verified things work as expected?

+1 from me if so

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java (hunk 1 above)
Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java:122]: I think this is intentional, we want to reduce the request to namenode...

**Comment by Apache9** [comment]
I think the key problem here is we use StoreFileInfo as the key of a hash map.

In general, within a typical store, we just need to use the storefile's name to decide whether they are the same StoreFileInfo.

Or at least, when comparing, we should normalize the Path object to the same format...

**Comment by Apache9** [comment]
OK, there is a FileSystem instance in the StoreFileInfo, so I think we just need to make a makeQualified to normalize the initial path?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java (hunk 1 above)
Comment (wchevreuil) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java:122]: Ok, haven't thought about that. Let me revert this, and sanitize the path, instead.

**Comment by wchevreuil** [comment]
> I assume you've spot-checked the output of TestHStore after this and verified things work as expected?
> 
> +1 from me if so

Yep, TestHStore was fine, but noticed TestStoreFileInfo failed here. Need to double check on that.

**Review by Apache9 — APPROVED** [review]

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 17s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 26s |  master passed  |
| +1 :green_heart: |  compile  |   2m 16s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 32s |  master passed  |
| +1 :green_heart: |  spotless  |   0m 43s |  branch has no errors when running spotless:check.  |
| +1 :green_heart: |  spotbugs  |   1m 23s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 13s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 13s |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 13s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 30s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  11m 34s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.2 3.3.1.  |
| -1 :x: |  spotless  |   0m 35s |  patch has 24 errors when running spotless:check, run spotless:apply to fix.  |
| +1 :green_heart: |  spotbugs  |   1m 22s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 11s |  The patch does not generate ASF License warnings.  |
|  |   |  32m 19s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4419 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti spotless checkstyle compile |
| uname | Linux 44e1dc709ff6 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / a712f9fa34 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 64 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 45s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 21s |  master passed  |
| +1 :green_heart: |  compile  |   1m  3s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 50s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 34s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   4m 27s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 26s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 218m 57s |  hbase-server in the patch passed.  |
|  |   | 240m 47s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4419 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux a1ee010400a9 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / a712f9fa34 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/testReport/ |
| Max. process+thread count | 2386 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4419/2/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by joshelser — APPROVED** [review]
New patch looks good too.

**Pull request merged**